corporate/views.py

```diff
@@ -159,12 +179,23 @@ def initial_upgrade(request: HttpRequest) -> HttpResponse:
         return render(request, "404.html", status=404)
 
     billing_page_url = reverse(billing_home)
+    if request.GET.get("onboarding") is not None:
+        billing_page_url = f"{billing_page_url}?onboarding=true"
+
+    realm_on_free_trial = False
 
     customer = get_customer_by_realm(user.realm)
-    if customer is not None and (get_current_plan_by_customer(customer) is not None or customer.sponsorship_pending):
-        if request.GET.get("onboarding") is not None:
-            billing_page_url = f"{billing_page_url}?onboarding=true"
-        return HttpResponseRedirect(billing_page_url)
+    if customer:
+        plan = get_current_plan_by_customer(customer)
+        if plan:
+            if plan.is_free_trial():
+                if not plan.charge_automatically or customer_has_credit_card_as_default_source(customer):
+                    return HttpResponseRedirect(billing_page_url)
+                realm_on_free_trial = True
+            else:
```

Comment:
Does this take into account the case of `CustomerPlan.ENDED`? Is it possible to do an upgrade to restart an ended plan?

Reply:
Though reading the code that's being replaced, I see that we used to do the same in the ENDED case anyway, so I assume this is correct.

Reply:
Yeah. The customer can restart the plan. But that would start the plan again in FREE TRIAL. We should add a check to not give free trial again for users who have tried it once. I would add that once this PR is merged.